sonar-core/src/main/java/org/sonar/core/platform/PluginLoader.java

```diff
@@ -45,117 +38,98 @@
+        Version minSqVersion = info.getMinimalSqVersion();
+        boolean compatibilityMode = (minSqVersion != null && minSqVersion.compareToIgnoreQualifier(COMPATIBILITY_MODE_MAX_VERSION) < 0);
+        def.setCompatibilityMode(compatibilityMode);
+        if (compatibilityMode) {
+          Loggers.get(getClass()).info("API compatibility mode is enabled on plugin {} [{}] " +
+            "(built with API lower than {})",
+            info.getName(), info.getKey(), COMPATIBILITY_MODE_MAX_VERSION);
         }
       }
     }
-    Map<String, ClassLoader> classloadersByBasePluginKey = builder.build();
-    for (ClassloaderDef def : defs) {
-      ClassLoader builtClassloader = classloadersByBasePluginKey.get(def.getBasePluginKey());
-      if (builtClassloader == null) {
-        throw new IllegalStateException(String.format("Fail to create classloader for plugin [%s]", def.getBasePluginKey()));
-      }
-      def.setBuiltClassloader(builtClassloader);
-    }
+    return classloadersByBasePlugin.values();
   }
 
   /**
-   * Step 3 - instantiate plugin instances ({@link Plugin}
+   * Instantiates collection of ({@link Plugin} according to given metadata and classloaders
    *
    * @return the instances grouped by plugin key
    * @throws IllegalStateException if at least one plugin can't be correctly loaded
    */
-  private Map<String, Plugin> instantiatePluginInstances(Collection<ClassloaderDef> defs) {
+   @VisibleForTesting
+   Map<String, Plugin> instantiatePluginClasses(Map<PluginClassloaderDef, ClassLoader> classloaders) {
     // instantiate plugins
     Map<String, Plugin> instancesByPluginKey = new HashMap<>();
-    for (ClassloaderDef def : defs) {
+    for (Map.Entry<PluginClassloaderDef, ClassLoader> entry : classloaders.entrySet()) {
+      PluginClassloaderDef def = entry.getKey();
+      ClassLoader classLoader = entry.getValue();
+
       // the same classloader can be used by multiple plugins
-      for (Map.Entry<String, String> entry : def.getMainClassesByPluginKey().entrySet()) {
-        String pluginKey = entry.getKey();
-        String mainClass = entry.getValue();
+      for (Map.Entry<String, String> mainClassEntry : def.getMainClassesByPluginKey().entrySet()) {
+        String pluginKey = mainClassEntry.getKey();
+        String mainClass = mainClassEntry.getValue();
         try {
-          instancesByPluginKey.put(pluginKey, (Plugin) def.getBuiltClassloader().loadClass(mainClass).newInstance());
+          instancesByPluginKey.put(pluginKey, (Plugin) classLoader.loadClass(mainClass).newInstance());
         } catch (UnsupportedClassVersionError e) {
           throw new IllegalStateException(String.format("The plugin [%s] does not support Java %s",
             pluginKey, SystemUtils.JAVA_VERSION_TRIMMED), e);
-        } catch (ClassNotFoundException | IllegalAccessException | InstantiationException e) {
+        } catch (Throwable e) {
```

Comment:
![BLOCKER](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png) Catch Exception instead of Throwable. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1181)